src/test/runner.ts

```diff
@@ -230,6 +230,43 @@ export class Runner {
 
       await Promise.all(Array.from(outputDirs).map(outputDir => removeFolderAsync(outputDir).catch(e => {})));
 
+      let testGroups = createTestGroups(rootSuite);
+
+      const shard = config.shard;
+      if (shard) {
+        const shardGroups: TestGroup[] = [];
+        const shardTests = new Set<Test>();
+
+        // Each shard gets some tests.
+        const shardSize = Math.floor(total / shard.total);
+        // First few shards get one more test each.
+        const extraOne = total - shardSize * shard.total;
+
+        const from = shardSize * shard.current + Math.min(extraOne, shard.current);
+        const to = from + shardSize + (shard.current < extraOne ? 1 : 0);
+        let current = 0;
+        for (const group of testGroups) {
+          // Any test group goes to the shard that contains the first test of this group.
+          // So, this shard gets any group that starts at [from; to)
+          if (current >= from && current < to) {
+            shardGroups.push(group);
+            for (const test of group.tests)
+              shardTests.add(test);
+          }
+          current += group.tests.length;
+        }
+
+        testGroups = shardGroups;
+        filterSuite(rootSuite, () => false, test => shardTests.has(test));
+      }
+
+      if (process.stdout.isTTY) {
```

Comment:
I saw this logic was also there before, but would it not be even more relevant on the bots to see if shards are actively used and which one is used exactly?

Reply:
@mxschmitt: we can only touch console in reporters, library can't touch it. In this case, we make an exception, but ensure live user.